sublime_jedi/utils.py

```diff
@@ -1,169 +1,81 @@
-            callback(window.active_view(), response)
-            break
-
-
-class BaseThread(threading.Thread):
-
-    def __init__(self, fd, window_id, waiting, lock):
-        self.fd = fd
-        self.done = False
-        self.waiting = waiting
-        self.wait_lock = lock
-        self.window_id = window_id
-        super(BaseThread, self).__init__()
-        self.daemon = True
-        self.start()
-
-
-class ThreadReader(BaseThread):
-
-    def run(self):
-        while not self.done:
-            line = self.fd.readline()
-            if line:
-                data = None
-                try:
-                    data = json.loads(line.strip())
-                except ValueError:
-                    if not isinstance(data, dict):
-                        logger.exception(
-                            "Non JSON data from daemon: {0}".format(line)
-                        )
-                else:
-                    self.call_callback(data)
-
-    def call_callback(self, data):
-        """
-        Call callback for response data
+import jedi  # noqa
+from jedi.api.environment import Environment  # noqa
 
-        :type data: dict
-        """
-        if 'logging' in data:
-            getattr(logger, data['logging'])(data['content'])
-            return
 
-        with self.wait_lock:
-            callback = self.waiting.pop(data['uuid'], None)
-
-        if callback is not None:
-            delayed_callback = partial(
-                run_in_active_view,
-                self.window_id,
-                callback,
-                data[data['type']]
-            )
-            sublime.set_timeout(delayed_callback, 0)
+def ask_daemon(view, callback, ask_type, location=None):
+    """
+    Daemon request shortcut
 
+    :type view: sublime.View
+    :type callback: callabel
+    :type ask_type: str
+    :type location: type of (int, int) or None
+    """
+    window_id = view.window().id()
 
-class ThreadWriter(BaseThread, Queue):
+    if window_id not in DAEMONS:
+        DAEMONS[window_id] = Daemon(view)
 
-    def __init__(self, *args, **kwargs):
-        Queue.__init__(self)
-        super(ThreadWriter, self).__init__(*args, **kwargs)
+    def summon_daemon():
+        DAEMONS[window_id].request(view, ask_type, callback, location)
 
-    def run(self):
-        while not self.done:
-            request_data = self.get()
+    if PY3:
+        sublime.set_timeout_async(summon_daemon, 0)
+    else:
+        sublime.set_timeout(summon_daemon, 0)
 
-            if not request_data:
-                continue
 
-            callback, data = request_data
+class Daemon(object):
 
-            with self.wait_lock:
-                self.waiting[data['uuid']] = callback
+    def __init__(self, view):
+        self.window_id = view.window().id()
+        self.process = self._start_process(get_settings(view))
 
-            if not isinstance(data, str):
-                data = json.dumps(data)
+    def _start_process(self, settings):
```

Comment:
I think we no more start process here, looks like daemon configuration

Reply:
Yes, will rename it.